### ISSUE TITLE
Write a roadmap for kwik, and document how it will differ from Kotest

### ISSUE THREAD

**Issue by jcornaz** [jcornaz/kwik#175]
When I created this project, [kotest](https://github.com/kotest/kotest) was named "kotlintest" and its property based testing api was very different. I used it a while before deciding that it was not good enough for my taste and that I wanted to put efforts in creating my own property based testing library.

But the property based testing API of kotlintest is no more, and Kotest got a completely reworked new API which looks much closer to my preferences (and to what `kwik` exposes).

I have to evaluate `kotest` property based testing capabilities to answer the following questions:

**Is kwik a relevant alternative to kotest?**

If yes, then I should make sure I focus on the right parts of the API to offer a good* alternative to kotest.

\* "good" does not mean "better". `kwik` and `kotest` can be both excellent and yet *different*.

If kwik is not a relevant alternative to kotest, maybe my time would be better invested in contributing to `kotest` rather than continuing on `kwik`.

**Can kwik and kotest be *complementary*?**

Yes, the problem being solved is definitely the same. But maybe the two solutions are, or can be, *complementary*.

In which case, I should focus on the strength of `kwik` and try to offer a compatibility layer.

**Comment by jcornaz** [comment]
> Is kwik a relevant alternative to kotest?

Yes.

About the current API. `kwik` focuses more on generator combination and transformation than kotest, which relies more on creating generator from scratch, eventually by delegating to existing generators.

And the new experimental API I've started and which is more inspired by the property based testing available in elm and rust, takes a quite different approach than `Kotest`. My main goal with this new API is to reduce the number of problem detected when the tests are run, for more compile-time checks.

I don't think one approach is objectively better than the other. So people ought to choose between `Kotest` and `kwik`.

In short I would describe: 

* `Kotest` as more *intuitive* than how will `kwik` look like
* `kwik` as *safer* (catching more at compile time) than `Kotest`


I compatibility layer is also definitely an option. One could then profit from the `kwik` API powered by `Kotest` variety of generators. (But this is not a priority at the moment)

I let this issue open, and will close it when committing a roadmap going in more detail of the future for kwik.
